akka-http/src/main/scala/akka/http/scaladsl/server/RouteConcatenation.scala

```diff
@@ -34,12 +34,12 @@ trait RouteConcatenation {
 
 object RouteConcatenation extends RouteConcatenation {
 
-  class RouteWithConcatenation(route: Route) {
+  class RouteWithConcatenation(val route: Route) extends AnyVal {
     /**
      * Returns a Route that chains two Routes. If the first Route rejects the request the second route is given a
      * chance to act upon the request.
      */
-    def ~(other: Route): Route = { ctx ⇒
+    def ~(other: ⇒ Route): Route = { ctx ⇒
```

Comment:
wait wait wait... :-)

This causes quite a lot of new allocations in each and every route combination. Before merging such change it would have to be made sure that it does not terribly degrade performance.

Also, not sure if the binary compatibility ignores applies hare can actually be made safely.

I would not want to merge this change at this stage of the project.
If anything, an alternate combinator could be offered.

Reply:
@ktoso @johanandren I implemented a small performance test that compares 4 variants of concat operators (with/without value class, with/without by name argument). Preliminary results can be found here: https://github.com/swachter/akka-route-concat 

For sure, the measurements should be run with more iterations and forks. I will do this as soon as time permits. The results indicate that the concatenation operator that is implemented using a value class and uses a by-name argument for the second route has the highest throughput.

Reply:
Note that it is not only about throughput but also about memory pressure, which may not show up in the throughput of a micro bench but will make life miserable in general in a larger app. So make sure to look at those numbers from JMH a swell.

Reply:
@johanandren Can you please tell me how to measure memory pressure. (I did not find a corresponding sbt-jmh option.)

Reply:
That'd be `-prof gc`. Also as you said, please run with warmup enough that you see the result triggering JIT and stabilize (5-10 at least I'd guess), and then more iterations.

Reply:
@johanandren Thanks for the information. I ran the measurements with the following args: `sbt> jmh:run -i 20 -wi 20 -f 5 -prof gc` and update the results (cf. https://github.com/swachter/akka-route-concat). For convenience I repeat my summary here:

The concatenation combinator that is implemented by a value class and whose 2nd argument is by-name (i.e. route1) has the highest throughput and the smallest normalized allocation rate for all considered structures of routing trees. The difference in performance between the various concatenation operators gets more pronounced for deeper and broader routing trees.

If you agree with this result then I propose to add a new route concatenation operator (because of binary compatibility) that is implemented by a value class and has a by-name second argument.

Directives can be combined with the `|` combinator. In order to align routes with directives I would also offer the `|` symbol for that new combinator. An alias for the new combinator could be `or`.

Reply:
I basically agree with performance and the laziness argument. One change between spray and akka-http was that we chose to embrace the laziness in the routing tree even more and made `Directive0.apply` lazy (because that's what most people tripped over). Following up on that making even more parts of the routing structure lazy seems just consistent.

I'm actually not particularly concerned about performance (though testing it definitely makes sense). Routing is a high-level feature and we know that it is not super fast. The directive structure is very flexible but it doesn't come for free. In fact, right now the most expensive part is not the routing but the marshalling and content type negotiation code, so that's where you would have to start optimizing to improve performance.

On the other hand, by now, we usually recommend `concat` instead of `~`. The problem is that we cannot do the same change for `concat` because there are no repeated by-name parameters (`=> Route*`). So, we would either have to introduce another inconsistency or would have to provide overloads of `concat` for different arities.

There's another argument against doing anything and that is about why adding this laziness would actually make a large difference. As I said before `Directive0.apply` is now lazy. That means the only place where adding laziness to `~` is where *creating* the directive passed to `~` is expensive or evaluated unexpectedly eagerly. @swachter, do you have any examples of where that would be the case?